Core/include/Acts/Propagator/DenseEnvironmentExtension.hpp

```diff
@@ -350,18 +363,21 @@ struct DenseEnvironmentExtension {
   /// @param [in] state Deliverer of configurations
   template <typename propagator_state_t>
   void initializeEnergyLoss(const propagator_state_t& state) {
-    energy[0] = std::hypot(initialMomentum, state.options.mass);
+    using std::hypot;
+    energy[0] = hypot(initialMomentum, state.options.mass);
     // use unit length as thickness to compute the energy loss per unit length
     Acts::MaterialSlab slab(material, 1);
     // Use the same energy loss throughout the step.
     if (state.options.meanEnergyLoss) {
       g = -computeEnergyLossMean(slab, state.options.absPdgCode,
-                                 state.options.mass, qop[0]);
+                                 state.options.mass,
+                                 static_cast<double>(qop[0]));
```

Comment:
I think this is not possible, because of the conversation rules. `autodiff::dual<T>` does support a (implicit or explicit) conversion to `T`, which is `double` in our case. A `static_cast<float>(...)` would fail. So we must apply a `static_cast<double>` as a generic method to get an ordinary floating point number from `double` as well as `autodiff::dual<double>` which are the two use cases we have. When not wanting an implicit conversion from `double` to `float`, we would need `static_cast<float>(static_cast<double>(...))`.

Another method would be a custom metafunction `scalar_to_float` which has spezialisations for `double` and `autodiff::dual<double>`.

just out of interest: Is there a specific reason why the energy loss computations are done with `float`?

Reply:
Thanks, I was suspecting something like this. I agree with you that `static_cast<double>(...)` is the simplest solution. Let's leave it like that.

All material interactions are done in `float` since there really is no need for higher precision. The material parameters are only stored in `float` both to keep them the size down and because they are usually not known with high precision anyways. E.g. the energy loss is just the mean of a statistical process and the statistical variations are much larger than the float precision. Computing them in double would be a waste of resources.